tests/test_visualise_results.py

```diff
@@ -3,60 +3,160 @@
-    results = get_results(
-        folders=[
-            "./tests/mock_stats/2024-02-07T03:09:41",
-            "./tests/mock_stats/2024-02-06T03:09:41",
-        ]
-    )
-    dataframe = get_data_frame(results)
-    assert_frame_equal(dataframe, expected_data_frame_multiple_files)
 
+@pytest.mark.parametrize(
+    "folders, data_frame_method, expected_result",
+    (
+        (
+            [
+                "./tests/mock_stats/2024-02-07T03:09:41",
+                "./tests/mock_stats/2024-02-06T03:09:41",
+            ],
+            get_performance_data_frame,
+            expected_data_frame_multiple_files,
+        ),
+        (
+            [
+                "./tests/mock_stats/2024-07-25T03:09:41",
+                "./tests/mock_stats/2024-07-29T03:09:41",
+            ],
+            get_additional_metrics_data_frame,
+            expected_data_frame_multiple_files_session_pdf,
+        ),
+    ),
+)
+def test_get_data_frame_multiple_files(folders, data_frame_method, expected_result):
+    results = get_results(folders=folders)
+    dataframe = data_frame_method(results)
+    assert_frame_equal(dataframe, expected_result)
 
-def test_plot_data_df(mocker, get_results_single_file):
-    dataframe = get_data_frame(get_results_single_file)
+
+@pytest.mark.parametrize(
+    "expected_dataframe, data_frame_method, results_file",
+    (
+        (expected_data_frame, get_performance_data_frame, "get_results_single_file"),
+        (
+            expected_data_frame_session_pdf,
+            get_additional_metrics_data_frame,
+            "get_results_single_file_with_pdf_endpoint",
+        ),
+    ),
+)
+def test_plot_data_df(
+    expected_dataframe, data_frame_method, results_file, mocker, request
+):
+    dataframe = data_frame_method(request.getfixturevalue(results_file))
 
     mock_plot_data = mocker.patch("scripts.visualise_results.plot_data")
     mock_plot_data(dataframe, 1)
 
     assert mock_plot_data.call_count == 1
-    assert_frame_equal(mock_plot_data.call_args[0][0], expected_data_frame)
+    assert_frame_equal(mock_plot_data.call_args[0][0], expected_dataframe)
 
 
-def test_plot_data(mocker, get_results_single_file):
-    dataframe = get_data_frame(get_results_single_file)
+@pytest.mark.parametrize(
+    "results_file, data_frame_method, image_name",
+    (
+        (
+            "get_results_single_file",
+            get_performance_data_frame,
+            "performance_graph.png",
+        ),
+        (
+            "get_results_single_file_with_pdf_endpoint",
+            get_additional_metrics_data_frame,
+            "additional_metrics.png",
+        ),
+    ),
+)
+def test_individual_graph_creation(
+    results_file, data_frame_method, image_name, mocker, request
+):
+    dataframe = data_frame_method(request.getfixturevalue(results_file))
     try:
         graph_output = mocker.patch("matplotlib.pyplot.savefig")
-        plot_data(dataframe, 1)
+        create_graph([dataframe], 1, image_name)
```

Comment:
The behaviour is a bit weird now that even if you only attempt to plot 1 graph, the output image will still have 2 🤔

Reply:
The current logic creates two graphs because of the subplots. We'll have to keep subplots in order to have 2 graphs. Currently, for this test, we only intend to create "one graph", but there will be 2 subplots, the other subplot isn't populated with the data simply because there's no dataframe for the second subplot, so I guess this test may need to be changed or deleted because there cannot be only one graph where 2 subplots are used.

Reply:
Yeah similarly if you pass more than 2 dataframes to the method it breaks at the moment. This test is hinting that we are can handle a single dataframe when the method has been designed to render the same two dataframes each time.

Reply:
I think the code either needs to be clear that this method plots the two expected dataframes, or it needs to be made generic so if we pass 1 or multiple dataframes, it plots whatever is there.

Reply:
Updated 👍